meta-mbl-distro/classes/mbl-partitions.bbclass

```diff
@@ -24,138 +24,405 @@ inherit mbl-artifact-names
+            return p
+
+        def _set_offset_vars(self):
+            offset_vars = [ "MBL_{}_OFFSET1_KiB".format(self.name) ]
+            if self.is_banked:
+                offset_vars.append("MBL_{}_OFFSET2_KiB".format(self.name))
+
+            prev_offset = None
+            prev_size = None
+            prev_non_skipped_part = self._prev_non_skipped_part()
+            if prev_non_skipped_part:
+                prev_offset = prev_non_skipped_part.offsets_KiB[-1]
+                prev_size = prev_non_skipped_part.size_KiB
+
+            for offset_var in offset_vars:
+                offset = self._set_offset_var(offset_var, prev_offset, prev_size)
+                prev_offset = offset
+                prev_size = self.size_KiB
+
+        def _set_offset_var(self, offset_var, prev_offset, prev_size):
+            offset = None
+            have_offset = self._is_var_set(offset_var)
+            no_default = self._get_bool_var("MBL_{}_NO_DEFAULT_OFFSET".format(self.name))
+            have_prev = prev_offset is not None and prev_size is not None
+            if have_offset:
+                offset = self._get_size_var_or_fatal(offset_var)
+                self.offsets_KiB.append(offset)
+                if offset % self.align_KiB:
+                    bb.fatal("{}={}KiB is badly aligned. Expected alignment is {}KiB".format(
+                        offset_var, offset, self.align_KiB
+                    ))
+                if have_prev and prev_offset + prev_size > offset:
+                    bb.fatal("{}={}KiB overlaps previous partition with offset {}KiB and size {}KiB".format(
+                        offset_var, offset, prev_offset, prev_size
+                    ))
+            elif no_default:
+                bb.fatal("{} is not specified".format(offset_var))
+            elif have_prev:
+                offset = self._calculate_next_offset(prev_offset, prev_size)
+                self.d.setVar(offset_var, str(offset))
+                self.offsets_KiB.append(offset)
+            else:
+                bb.fatal("{} is not set", offset_var)
+            return offset
+
+        def _calculate_next_offset(self, prev_offset, prev_size):
+            last_byte_of_prev = prev_offset + prev_size - 1
+            return last_byte_of_prev + self.align_KiB - (last_byte_of_prev % self.align_KiB)
+
+        def _get_size_var_or_fatal(self, var):
+            var_str = self.d.getVar(var, True)
+            if var_str is None:
+                bb.fatal("{} has not been set".format(var))
+                return 0
+            if not var_str.isdigit():
+                bb.fatal("{} (\"{}\") must contain only decimal digits".format(var, var_str))
+                return 0
+            return int(var_str)
+
+        def _is_var_set(self, var):
+            return bool(self.d.getVar(var, True))
+
+        def _get_bool_var(self, var):
+            val = self.d.getVar(var, True)
+            return val is not None and bool(int(val))
```

Comment:
`bool(int(val))` could do with some explanation as it looks a bit strange to me. Should probably add explanatory docstrings to each method of this class. Some comments in the larger methods describing what is going on would also be useful.

Reply:
> Should probably add explanatory docstrings to each method of this class

Yep, that's on my list of things to do.

> `bool(int(val))`

Thanks, I'll make sure I explain this (or change it if I can no longer justify it to myself!).

Reply:
I've added docstrings and in-line comments in https://github.com/ARMmbed/meta-mbl/pull/617/commits/34e632c489ddce4f816e5f19992c7b47b7c6432a.

Reply:
Thanks, I've been trying to think of a way to drop this `bool(int(val))` but I can't really think of a good reason. `int(" ")` will throw an exception though so you may want to handle that.

Reply:
I'm not too bothered about it throwing an exception. I just want to avoid it silently doing the wrong thing.